contracts/programs/bubblegum/src/lib.rs

```diff
@@ -351,62 +551,80 @@ pub mod bubblegum {
+        assert_enough_mints_to_approve(&authority, request.mint_capacity)?;
+        authority.approve_mint_capacity(request.mint_capacity);
+        request.set_approved();
+
+        Ok(())
+    }
+
+    pub fn close_mint_request(ctx: Context<CloseMintRequest>) -> Result<()> {
+        let authority = &mut ctx.accounts.tree_authority;
+        let mint_authority = &mut ctx.accounts.mint_authority;
+        let request = &mut ctx.accounts.mint_authority_request;
+        let request_info = request.to_account_info();
+
+        // Transfer remaining mint capacity to authority
+        if request.is_approved() {
+            authority.restore_mint_capacity(request.mint_capacity);
+        }
+
+        // Transfer lamports
+        **mint_authority.lamports.borrow_mut() = mint_authority
+            .lamports()
+            .checked_add(request_info.lamports())
+            .ok_or(BubblegumError::CloseMintRequestError)?;
+        **request_info.lamports.borrow_mut() = 0;
+        Ok(())
+    }
+
+    pub fn set_tree_delegate(ctx: Context<SetTreeDelegate>) -> Result<()> {
+        ctx.accounts.tree_authority.delegate = ctx.accounts.new_delegate.key();
+        Ok(())
+    }
+
     pub fn mint_v1(ctx: Context<MintV1>, message: MetadataArgs) -> Result<()> {
         // TODO -> Pass collection in check collection authority or collection delegate authority signer
         // TODO -> Separate V1 / V1 into seperate instructions
-        assert_metadata_is_mpl_compatible(&message)?;
         let owner = ctx.accounts.owner.key();
         let delegate = ctx.accounts.delegate.key();
-        let merkle_slab = ctx.accounts.merkle_slab.to_account_info();
+        let merkle_slab = &ctx.accounts.merkle_slab;
 
-        // @dev: seller_fee_basis points is encoded twice so that it can be passed to marketplace instructions, without passing the entire, un-hashed MetadataArgs struct
-        let metadata_args_hash = keccak::hashv(&[message.try_to_vec()?.as_slice()]);
-        let data_hash = keccak::hashv(&[
-            &metadata_args_hash.to_bytes(),
-            &message.seller_fee_basis_points.to_le_bytes(),
-        ]);
-        let nonce = &mut ctx.accounts.authority;
-        let creator_data = message
-            .creators
-            .iter()
-            //TODO include verified
-            .map(|c| [c.address.as_ref(), &[c.share]].concat())
-            .collect::<Vec<_>>();
-        let creator_hash = keccak::hashv(
-            creator_data
-                .iter()
-                .map(|c| c.as_slice())
-                .collect::<Vec<&[u8]>>()
-                .as_ref(),
-        );
-        let asset_id = get_asset_id(&merkle_slab.key(), nonce.count);
-        let leaf = LeafSchema::new_v0(
-            asset_id,
+        let mint_authority = &ctx.accounts.mint_authority;
+        let authority = &mut ctx.accounts.authority;
+
+        if *mint_authority.key != authority.creator && *mint_authority.key != authority.delegate {
```

Comment:
So this kinda messes up the state because the `authority` and `mint_authority` can reduce the total capacity to the point where the requested mints might be unable to append

Reply:
1. Only the tree `creator` or `delegate` can mint freely and mess up the state. We can fix this by forcing them to use MintRequest's to mint as well.

2. MintRequests can be closed, so invalid state does not lock up lamports indefinitely.